File: src/bin/cli/cmd/mass_rename.rs
use crate::cli::OutputFormat;
use crate::dto::{MassRenameResultDto, RenameOperationDto};
use crate::json_output::{output_error, output_success};
use lib::backup::ports::BackupService;
use lib::backup::service::SubRipBackupService;
use lib::rename::model::RenameError;
use lib::rename::ports::{RenameService, TemplateContext};
use lib::rename::service::FileRenameService;
use log::{debug, error, info};
use std::io::{self, Write};

#[allow(clippy::too_many_arguments)]
pub fn handle(
    file_mask: &str,
    dry_run: bool,
    force: bool,
    series_mode: bool,
    name: Option<String>,
    season: Option<u32>,
    language: Option<String>,
    separator: &str,
    file_template: &str,
    format: &OutputFormat,
) -> anyhow::Result<()> {
    info!("starting mass-rename with mask: {}", file_mask);
    debug!(
        "dry_run: {}, force: {}, series_mode: {}",
        dry_run, force, series_mode
    );

    let current_dir = std::env::current_dir()?;
    let service = FileRenameService::new(current_dir.clone());

    let files = match service.find_files(file_mask) {
        Ok(files) => files,
        Err(e) => {
            let (code, msg) = format_rename_error(&e);
            output_error(format, code, msg, None);
            return Err(anyhow::anyhow!(""));
        }
    };

    info!("found {} files", files.len());

    let context = TemplateContext::new()
        .with_name(name)
        .with_season(season)
        .with_language(language)
        .with_separator(Some(separator.to_string()));

    let operations =
        match service.prepare_rename_operations(files, file_template, &context, series_mode) {
            Ok(ops) => ops,
            Err(e) => {
                let (code, msg) = format_rename_error(&e);
                output_error(format, code, msg, None);
                return Err(anyhow::anyhow!(""));
            }
        };

    let op_dtos: Vec<RenameOperationDto> = operations
        .iter()
        .map(|op| RenameOperationDto {
            original: op
                .original_filename()
                .unwrap_or_else(|| "unknown".to_string()),
            new_name: op.new_name.clone(),
            collision: op.collision,
        })
        .collect();

    if dry_run {
        info!("dry-run mode, operations not executed");

        let dto = MassRenameResultDto {
            renamed_count: 0,
            skipped_count: 0,
            operations: op_dtos,
            dry_run: true,
        };

        output_success(format, &dto, || {
            println!("Preview:");
            println!();
            for op in &operations {
                let original_name = op
                    .original_filename()
                    .unwrap_or_else(|| "unknown".to_string());
                println!("  {} -> {}", original_name, op.new_name);
                if op.collision {
                    println!("    Warning: file already exists");
                }
            }
            println!();
            println!("Dry-run mode: no files were renamed");
        });

        return Ok(());
    }

    // Show preview in text mode
    if matches!(format, OutputFormat::Text) {
        println!("Preview:");
        println!();
        for op in &operations {
            let original_name = op
                .original_filename()
                .unwrap_or_else(|| "unknown".to_string());
            println!("  {} -> {}", original_name, op.new_name);
            if op.collision {
                println!("    Warning: file already exists");
            }
        }
        println!();
    }

    if !force && matches!(format, OutputFormat::Text) {
        print!("Proceed with renaming {} files? (y/N): ", operations.len());
        io::stdout().flush()?;

        let mut input = String::new();
        io::stdin().read_line(&mut input)?;

        if !input.trim().eq_ignore_ascii_case("y") {
            info!("operation cancelled by user");
            println!("Cancelled");
            return Ok(());
        }
    }

    // Execute rename operations
    info!("starting file renaming");
    let backup_service = SubRipBackupService::new();
    let mut renamed_count = 0;
    let mut skipped_count = 0;

    for op in operations {
        let original_name = op
            .original_filename()
            .unwrap_or_else(|| "unknown".to_string());

        if op.collision && matches!(format, OutputFormat::Text) {
            print!("File '{}' already exists. Overwrite? (y/N): ", op.new_name);
            io::stdout().flush()?;

            let mut input = String::new();
            io::stdin().read_line(&mut input)?;

            if !input.trim().eq_ignore_ascii_case("y") {
                info!("skipping file due to collision: {:?}", op.original_path);
                println!("  Skipped: {}", original_name);
                skipped_count += 1;
                continue;
            }
        } else if op.collision {
            // In JSON mode, skip collisions without prompting
            skipped_count += 1;
            continue;
        }

        match backup_service.create_backup(&op.original_path) {
            Ok(Some(_)) => {}
            Ok(None) => {
                error!("file does not exist, cannot be renamed");
                if matches!(format, OutputFormat::Text) {
                    eprintln!("  Error: File not found: {}", original_name);
                }
                skipped_count += 1;
                continue;
            }
            Err(e) => {
                error!("failed to create backup: {}", e);
                if matches!(format, OutputFormat::Text) {
                    eprintln!(
                        "  Error: Failed to create backup for {}: {}",
                        original_name, e
                    );
                }
                skipped_count += 1;
                continue;
            }
        }

        let new_path = current_dir.join(&op.new_name);

        match std::fs::rename(&op.original_path, &new_path) {
            Ok(_) => {
                info!("file renamed: {} -> {}", original_name, op.new_name);
                if matches!(format, OutputFormat::Text) {
                    println!("  {} -> {}", original_name, op.new_name);
                }
                renamed_count += 1;
            }
            Err(e) => {
                error!("failed to rename file: {}", e);
                if matches!(format, OutputFormat::Text) {
                    eprintln!("  Failed to rename {}: {}", original_name, e);
                }
                skipped_count += 1;
            }
        }
    }

    let dto = MassRenameResultDto {
        renamed_count,
        skipped_count,
        operations: op_dtos,
        dry_run: false,
    };

    output_success(format, &dto, || {
        println!();
        println!("Summary:");
        println!(
            "  {} file{} renamed",
            renamed_count,
            if renamed_count == 1 { "" } else { "s" }
        );
        if skipped_count > 0 {
            println!(
                "  {} file{} skipped",
                skipped_count,
                if skipped_count == 1 { "" } else { "s" }
            );
        }
    });

    Ok(())
}

fn format_rename_error(e: &RenameError) -> (&str, &str) {
    match e {
        RenameError::NoFilesFound(_) => ("no_files_found", "No files found matching pattern"),
        RenameError::TemplateError(_) => ("template_error", "Template rendering failed"),
        RenameError::IoError(_) => ("io_error", "File operation failed"),
        RenameError::InvalidTemplate(_) => ("invalid_template", "Invalid template"),
        RenameError::GlobError(_) => ("glob_error", "Glob pattern error"),
    }
}

File: internal/cli/cmd/mass_rename.go
package cmd

import (
	"bufio"
	"errors"
	"fmt"
	"log/slog"
	"os"
	"path/filepath"
	"strings"

	"github.com/lebe-dev/submarine/internal/cli"
	"github.com/lebe-dev/submarine/pkg/backup"
	"github.com/lebe-dev/submarine/pkg/rename"
)

// HandleMassRename is a 1-to-1 port of the Rust `cmd::mass_rename::handle`.
func HandleMassRename(
	fileMask string,
	dryRun bool,
	force bool,
	seriesMode bool,
	name *string,
	season *uint32,
	language *string,
	separator string,
	fileTemplate string,
	format cli.OutputFormat,
) error {
	slog.Info(fmt.Sprintf("starting mass-rename with mask: %s", fileMask))
	slog.Debug(fmt.Sprintf(
		"dry_run: %t, force: %t, series_mode: %t",
		dryRun, force, seriesMode,
	))

	currentDir, err := os.Getwd()
	if err != nil {
		return err
	}
	service := rename.NewFileRenameService(currentDir)

	files, findErr := service.FindFiles(fileMask)
	if findErr != nil {
		var re *rename.RenameError
		errors.As(findErr, &re)
		code, msg := formatRenameError(re)
		cli.OutputError(format, code, msg, nil)
		return errors.New("")
	}

	slog.Info(fmt.Sprintf("found %d files", len(files)))

	sep := separator
	context := rename.NewTemplateContext().
		WithName(name).
		WithSeason(season).
		WithLanguage(language).
		WithSeparator(&sep)

	operations, prepErr := service.PrepareRenameOperations(files, fileTemplate, &context, seriesMode)
	if prepErr != nil {
		var re *rename.RenameError
		errors.As(prepErr, &re)
		code, msg := formatRenameError(re)
		cli.OutputError(format, code, msg, nil)
		return errors.New("")
	}

	opDtos := make([]cli.RenameOperationDto, 0, len(operations))
	for _, op := range operations {
		original := renameOriginalName(op)
		opDtos = append(opDtos, cli.RenameOperationDto{
			Original:  original,
			NewName:   op.NewName,
			Collision: op.Collision,
		})
	}

	if dryRun {
		slog.Info("dry-run mode, operations not executed")

		dto := cli.MassRenameResultDto{
			RenamedCount: 0,
			SkippedCount: 0,
			Operations:   opDtos,
			DryRun:       true,
		}

		cli.OutputSuccess(format, dto, func() {
			fmt.Println("Preview:")
			fmt.Println()
			for _, op := range operations {
				originalName := renameOriginalName(op)
				fmt.Printf("  %s -> %s\n", originalName, op.NewName)
				if op.Collision {
					fmt.Println("    Warning: file already exists")
				}
			}
			fmt.Println()
			fmt.Println("Dry-run mode: no files were renamed")
		})

		return nil
	}

	// Show preview in text mode
	if format == cli.OutputFormatText {
		fmt.Println("Preview:")
		fmt.Println()
		for _, op := range operations {
			originalName := renameOriginalName(op)
			fmt.Printf("  %s -> %s\n", originalName, op.NewName)
			if op.Collision {
				fmt.Println("    Warning: file already exists")
			}
		}
		fmt.Println()
	}

	if !force && format == cli.OutputFormatText {
		fmt.Printf("Proceed with renaming %d files? (y/N): ", len(operations))
		if flushErr := massRenameStdoutFlush(); flushErr != nil {
			return flushErr
		}

		input, readErr := massRenameReadLine()
		if readErr != nil {
			return readErr
		}

		if !strings.EqualFold(strings.TrimSpace(input), "y") {
			slog.Info("operation cancelled by user")
			fmt.Println("Cancelled")
			return nil
		}
	}

	// Execute rename operations
	slog.Info("starting file renaming")
	backupService := backup.NewSubRipBackupService()
	renamedCount := 0
	skippedCount := 0

	for _, op := range operations {
		originalName := renameOriginalName(op)

		if op.Collision && format == cli.OutputFormatText {
			fmt.Printf("File '%s' already exists. Overwrite? (y/N): ", op.NewName)
			if flushErr := massRenameStdoutFlush(); flushErr != nil {
				return flushErr
			}

			input, readErr := massRenameReadLine()
			if readErr != nil {
				return readErr
			}

			if !strings.EqualFold(strings.TrimSpace(input), "y") {
				slog.Info(fmt.Sprintf("skipping file due to collision: %q", op.OriginalPath))
				fmt.Printf("  Skipped: %s\n", originalName)
				skippedCount++
				continue
			}
		} else if op.Collision {
			// In JSON mode, skip collisions without prompting
			skippedCount++
			continue
		}

		backupResult, backupErr := backupService.CreateBackup(op.OriginalPath)
		switch {
		case backupErr != nil:
			slog.Error(fmt.Sprintf("failed to create backup: %s", backupErr))
			if format == cli.OutputFormatText {
				fmt.Fprintf(os.Stderr,
					"  Error: Failed to create backup for %s: %s\n",
					originalName, backupErr,
				)
			}
			skippedCount++
			continue
		case backupResult == nil:
			slog.Error("file does not exist, cannot be renamed")
			if format == cli.OutputFormatText {
				fmt.Fprintf(os.Stderr, "  Error: File not found: %s\n", originalName)
			}
			skippedCount++
			continue
		}

		newPath := filepath.Join(currentDir, op.NewName)

		if renameErr := os.Rename(op.OriginalPath, newPath); renameErr != nil {
			slog.Error(fmt.Sprintf("failed to rename file: %s", renameErr))
			if format == cli.OutputFormatText {
				fmt.Fprintf(os.Stderr, "  Failed to rename %s: %s\n", originalName, renameErr)
			}
			skippedCount++
			continue
		}

		slog.Info(fmt.Sprintf("file renamed: %s -> %s", originalName, op.NewName))
		if format == cli.OutputFormatText {
			fmt.Printf("  %s -> %s\n", originalName, op.NewName)
		}
		renamedCount++
	}

	dto := cli.MassRenameResultDto{
		RenamedCount: renamedCount,
		SkippedCount: skippedCount,
		Operations:   opDtos,
		DryRun:       false,
	}

	cli.OutputSuccess(format, dto, func() {
		fmt.Println()
		fmt.Println("Summary:")
		fmt.Printf(
			"  %d file%s renamed\n",
			renamedCount,
			massRenamePlural(renamedCount),
		)
		if skippedCount > 0 {
			fmt.Printf(
				"  %d file%s skipped\n",
				skippedCount,
				massRenamePlural(skippedCount),
			)
		}
	})

	return nil
}

// renameOriginalName returns the original filename for an operation, defaulting
// to "unknown" (was `op.original_filename().unwrap_or_else(|| "unknown".to_string())`).
func renameOriginalName(op rename.RenameOperation) string {
	if name, ok := op.OriginalFilename(); ok {
		return name
	}
	return "unknown"
}

// formatRenameError is a 1-to-1 port of the Rust `format_rename_error`.
func formatRenameError(e *rename.RenameError) (string, string) {
	switch e.Kind {
	case rename.ErrNoFilesFound:
		return "no_files_found", "No files found matching pattern"
	case rename.ErrTemplateError:
		return "template_error", "Template rendering failed"
	case rename.ErrIoError:
		return "io_error", "File operation failed"
	case rename.ErrInvalidTemplate:
		return "invalid_template", "Invalid template"
	case rename.ErrGlobError:
		return "glob_error", "Glob pattern error"
	}
	return "io_error", "File operation failed"
}

// osStdoutFlush flushes stdout. Mirrors the Rust `io::stdout().flush()?`. The Go
// standard os.Stdout is unbuffered, so this is a no-op kept for parity.
func massRenameStdoutFlush() error {
	return nil
}

// massRenameStdin is a shared buffered reader over stdin, so that consecutive
// readLine calls do not lose buffered-but-unconsumed input (mirroring Rust's
// single locked io::stdin() handle).
var massRenameStdin = bufio.NewReader(os.Stdin)

// readLine reads a single line from stdin, mirroring the Rust
// `io::stdin().read_line(&mut input)?`.
func massRenameReadLine() (string, error) {
	line, err := massRenameStdin.ReadString('\n')
	if err != nil && line == "" {
		return "", err
	}
	return line, nil
}

// massRenamePlural returns "" when n == 1 and "s" otherwise. Mirrors the Rust
// inline `if count == 1 { "" } else { "s" }`.
func massRenamePlural(n int) string {
	if n == 1 {
		return ""
	}
	return "s"
}
